### PR TITLE
Exo question add fields

### PR DESCRIPTION
**Pull request by eldoniel**
This adds some fields we used in V5 at innovalangues.

The fields are set as follow:
- Functional instruction is set on the "description" field.
- Object of the question uses the ObjectQuestion entity, refering to a question and a resourceNode.
- Pedagogic instruction is set on the "specification" field.
- Complementary information is set on the "supplementary" field.

This PR doesn't update the editor in order to fill these fields, it only allows to show existing data in the player.

### CODE REVIEW

File: plugin/exo/Entity/Question.php [hunk 1]
```diff
@@ -119,7 +125,7 @@ public function setType($type)
         $this->type = $type;
     }
 
-    /**
+    /**, orphanRemoval=true
```
Comment (stefk) [inline plugin/exo/Entity/Question.php]: Should be removed.

File: plugin/exo/Manager/QuestionManager.php [hunk 2]
```diff
@@ -96,13 +102,32 @@ public function importQuestion(\stdClass $data)
...
+        $data->supplementary = $question->getSupplementary();
+        $data->specification = $question->getSpecification();
+        $data->objects = array_map(function ($object) use ($rm) {
+            $resourceObjectData = new \stdClass();
+            $resourceObjectData->id = (string) $object->getResourceNode()->getId();
+            $resourceObjectData->type = $object->getResourceNode()->getMimeType();
+            $resourceObjectData->name = $object->getResourceNode()->getName();
+            $resourceObjectData->path = $object->getResourceNode()->getPathForDisplay();
+            switch ($object->getResourceNode()->getMimeType()) {
+                case 'custom/hevinci_url':
```
Comment (stefk) [inline plugin/exo/Manager/QuestionManager.php]: Is this really necessary? Exo plugin shouldn't be aware of url plugin IMO.

File: plugin/exo/Entity/ObjectQuestion.php [hunk 3]
```diff
@@ -122,4 +122,20 @@ public function getType()
...
+     */
+    public function getResourceNode()
+    {
+        return $this->resourceNode;
+    }
+
+    /**
+     * @param ResourceNode $resourceNode
+     */
+    public function setResourceNode(ResourceNode $resourceNode)
```
Comment (Elorfin) [inline plugin/exo/Entity/ObjectQuestion.php]: use statement for `ResourceNode` is missing.

@stefk : why is this not catched by automatic analysis ? Isn't it part of the new config ?

File: plugin/exo/Manager/QuestionManager.php [hunk 4]
```diff
@@ -96,13 +102,32 @@ public function importQuestion(\stdClass $data)
...
         $data->title = $question->getTitle();
         $data->description = $question->getDescription();
         $data->invite = $question->getInvite();
+        $data->supplementary = $question->getSupplementary();
+        $data->specification = $question->getSpecification();
+        $data->objects = array_map(function ($object) use ($rm) {
+            $resourceObjectData = new \stdClass();
+            $resourceObjectData->id = (string) $object->getResourceNode()->getId();
+            $resourceObjectData->type = $object->getResourceNode()->getMimeType();
+            $resourceObjectData->name = $object->getResourceNode()->getName();
```
Comment (Elorfin) [inline plugin/exo/Manager/QuestionManager.php]: This is not needed and not part of the JSON Schema (see `Content` section).

File: plugin/exo/Manager/QuestionManager.php [hunk 5]
```diff
@@ -96,13 +102,32 @@ public function importQuestion(\stdClass $data)
...
+            $resourceObjectData->path = $object->getResourceNode()->getPathForDisplay();
```
Comment (Elorfin) [inline plugin/exo/Manager/QuestionManager.php]: Idem

File: plugin/exo/Manager/QuestionManager.php [hunk 6]
```diff
@@ -96,13 +102,32 @@ public function importQuestion(\stdClass $data)
...
+            switch ($object->getResourceNode()->getMimeType()) {
+                case 'custom/hevinci_url':
+                    $resourceObjectData->resourceUrl = $rm->getResourceFromNode($object->getResourceNode())->getUrl();
```
Comment (Elorfin) [inline plugin/exo/Manager/QuestionManager.php]: Property needs to be named `url` : http://json-quiz.github.io/json-quiz/spec/content.html.

File: plugin/exo/Manager/QuestionManager.php [hunk 7]
```diff
@@ -96,13 +102,32 @@ public function importQuestion(\stdClass $data)
...
+                    break;
+                default:
+                    $resourceObjectData->resourceUrl = $rm->getResourceFromNode($object->getResourceNode())->getHashName();
```
Comment (Elorfin) [inline plugin/exo/Manager/QuestionManager.php]: The default case needs to be the real URL to access the Resource (from memories, the route claro_resource_open).

**Comment by Elorfin** [comment]
You also need to manage the case of HTML resources which use an attribute `data` (instead of `url`) in exports.

File: plugin/exo/Resources/modules/question/Partials/show.html [hunk 8]
```diff
@@ -45,6 +45,10 @@ <h4 class="panel-title">
             <!-- Description of the Question -->
             <div data-ng-bind="questionShowCtrl.question.description"></div>
         </div>
+        <div class="question-info" data-ng-if="questionShowCtrl.question.objects.length > 0">
+            <h5>{{ 'question_object'|trans:{}:'ujm_sequence' }}</h5>
+            <div data-ng-repeat="object in questionShowCtrl.question.objects">{{ object.resourceUrl }}</div>
```
Comment (Elorfin) [inline plugin/exo/Resources/modules/question/Partials/show.html]: For content with `url` property, this needs to be a clickable link.

File: plugin/exo/Entity/ObjectQuestion.php (hunk 3 above)
Comment (stefk) [inline plugin/exo/Entity/ObjectQuestion.php]: It only detects uses of fully-qualified class names ( ≃ classes starting with `\`, without `use` statements). This isn't the case here. As far as the tool is concerned, `ResourceNode` is just a class in the `UJM\ExoBundle\Entity` namespace.

Comment (stefk) [inline plugin/exo/Entity/ObjectQuestion.php]: An automatic test would have caught it, tough...

File: plugin/exo/Resources/translations/ujm_sequence.fr.yml [hunk 9]
```diff
@@ -92,6 +92,9 @@ points: Points
 question_hints: Indices
 question_hint_label: Indice
 question_hint_show_title: Voir l'indice
+question_object: Objet de la question
+question_specification: Specification
+question_supplementary: Supplementary information
```
Comment (LaurentGruber) [inline plugin/exo/Resources/translations/ujm_sequence.fr.yml]: ça devrait être du FR

**Comment by LaurentGruber** [comment]
LGTM

**Comment by purplefish32** [comment]
LGTM

**Pull request merged**